admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,640 @@
+    joint_command_interface_[index].clear();
+    joint_state_interface_[index].clear();
+  }
+  release_interfaces();
+
+  force_torque_sensor_->release_interfaces();
+
+  return CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type AdmittanceController::update(
+  const rclcpp::Time & /*time*/, const rclcpp::Duration & period)
+{
+  if (admittance_->parameters_.enable_parameter_update_without_reactivation_)
+  {
+    admittance_->parameters_.update();
+  }
+
+  // get input commands
+  auto input_wrench_cmd = input_wrench_command_.readFromRT();
+  auto input_joint_cmd = input_joint_command_.readFromRT();
+  auto input_pose_cmd = input_pose_command_.readFromRT();
+
+  // Position has to always be there
+  auto num_joints = joint_state_interface_[0].size();
+  trajectory_msgs::msg::JointTrajectoryPoint current_joint_states;
+  current_joint_states.positions.resize(num_joints);
+  current_joint_states.velocities.resize(num_joints, 0.0);
+  current_joint_states.accelerations.resize(num_joints, 0.0);
+  trajectory_msgs::msg::JointTrajectoryPoint desired_joint_states;
+  desired_joint_states.positions.resize(num_joints);
+  desired_joint_states.velocities.resize(num_joints);
+  desired_joint_states.accelerations.resize(num_joints);
+
+  read_state_from_hardware(current_joint_states);
+
+  if (admittance_->parameters_.open_loop_control_) {
+    // TODO(destogl): This may not work in every case.
+    // Please add checking which states are available and which not!
+    current_joint_states = last_commanded_state_;
+  }
+
+  geometry_msgs::msg::Wrench ft_values;
+  force_torque_sensor_->get_values_as_message(ft_values);
+
+  // TODO(destogl): Enable this when unified mode is used
+//   if (admittance_->unified_mode_) {
+  //     admittance_->update(current_joint_states, ft_values, **input_pose_cmd, **input_wrench_cmd, period, desired_joint_states);
+//   } else {
+
+  // TODO(destogl): refactor this into different admittance controllers: 1. Pose input, Joint State input and Unified mode (is there need for switching between unified and non-unified mode?)
+  if (use_joint_commands_as_input_) {
+    std::array<double, 6> joint_deltas;
+    // If there are no positions, expect velocities
+    // TODO(destogl): add error handling
+    if ((*input_joint_cmd)->points[0].positions.empty()) {
+      for (auto index = 0u; index < num_joints; ++index) {
+        joint_deltas[index] = (*input_joint_cmd)->points[0].velocities[index] * period.seconds();
+      }
+    } else {
+      for (auto index = 0u; index < num_joints; ++index) {
+        // TODO(destogl): ATTENTION: This does not work properly, deltas are getting neutralized and robot is not moving on external forces
+        // TODO(anyone): Is here OK to use shortest_angular_distance?
+        joint_deltas[index] = angles::shortest_angular_distance(current_joint_states.positions[index], (*input_joint_cmd)->points[0].positions[index]);
```

Comment:
First time looking at this in awhile!

I just wanted to note, this should probably use shortest_angular_distance_with_large_limits().

https://github.com/ros/angles/blob/811da0230f13348fd40337f63219c3b1520088f3/angles/include/angles/angles.h#L222